flake8_trio.py

```diff
@@ -34,29 +33,116 @@ def is_trio_call(node: ast.AST, *names: str) -> Optional[str]:
 class Visitor(ast.NodeVisitor):
     def __init__(self) -> None:
         super().__init__()
-        self.problems: list[tuple[int, int]] = []
+        self.problems: List[Error] = []
+        self.safe_yields: Set[ast.Yield] = set()
 
     def visit_With(self, node: ast.With) -> None:
-        for item in node.items:
-            call = is_trio_call(item.context_expr, "fail_after", "move_on_after")
-            if call and not any(isinstance(x, ast.Await) for x in ast.walk(node)):
-                self.problems.append(
-                    (item.lineno, item.col_offset, TRIO100.format(call))
-                )
+        self.check_for_trio100(node)
+        self.check_for_trio101(node)
+        self.generic_visit(node)
+
+    def visit_AsyncWith(self, node: ast.AsyncWith) -> None:
+        self.check_for_trio100(node)
+        self.check_for_trio101(node)
+        self.generic_visit(node)
 
-        # Don't forget to visit the child nodes for other errors!
+    def visit_FunctionDef(self, node: ast.FunctionDef) -> None:
+        self.trio101_mark_yields_safe(node)
         self.generic_visit(node)
 
+    def visit_AsyncFunctionDef(self, node: ast.AsyncFunctionDef) -> None:
+        self.trio101_mark_yields_safe(node)
+        self.generic_visit(node)
+
+    def check_for_trio100(self, node: Union[ast.With, ast.AsyncWith]) -> None:
+        # Context manager with no `await` call within
+        for item in (i.context_expr for i in node.items):
+            call = is_trio_call(item, "fail_after", "move_on_after")
```

Comment:
[Looks like we should also check for `CancelScope`, `fail_at`, and `move_on_at`](https://trio.readthedocs.io/en/stable/reference-core.html#cancellation-api-details).

Reply:
hmm, `CancelScope` looks tricky to check for lexicographically - would have to parse out the value of everything that's `with`'d to see if it's a `CancelScope` defined somewhere else. But I can give it a try if you want

Reply:
Nah, I meant literally just checking for inline calls to CancelScope - `with trio.CancelScope() as scope:` is code that someone might write, and it's easy to check for *that* case.

Reply:
ah right, makes sense